dnsnameresolver.go

```diff
@@ -0,0 +1,291 @@
+package ocp_dnsnameresolver
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/coredns/coredns/plugin"
+
+	dnsv1alpha1 "github.com/openshift/api/network/v1alpha1"
+	networkclient "github.com/openshift/client-go/network/clientset/versioned"
+	networkclientv1alpha1 "github.com/openshift/client-go/network/clientset/versioned/typed/network/v1alpha1"
+	networkinformer "github.com/openshift/client-go/network/informers/externalversions"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+type namespaceDNSInfo map[string]string
+
+// OCPDNSNameResolver is a plugin that looks up responses from other plugins
+// and updates the status of DNSNameResolver objects.
+type OCPDNSNameResolver struct {
+	Next plugin.Handler
+
+	// configurable fields.
+	namespaces       map[string]struct{}
+	minimumTTL       int32
+	failureThreshold int32
+
+	// maps for storing regular and wildcard DNS name info.
+	// data mapping: DNS name --> Namespace --> DNSNameResolver object name
+	regularDNSInfo  map[string]namespaceDNSInfo
+	wildcardDNSInfo map[string]namespaceDNSInfo
+	regularMapLock  sync.Mutex
+	wildcardMapLock sync.Mutex
+
+	// client and informer for handling DNSNameResolver objects.
+	dnsNameResolverClient   networkclientv1alpha1.NetworkV1alpha1Interface
+	dnsNameResolverInformer cache.SharedIndexInformer
+	stopCh                  chan struct{}
+	stopLock                sync.Mutex
+	shutdown                bool
+}
+
+// New returns an initialized OCPDNSNameResolver with default settings.
+func New() *OCPDNSNameResolver {
+	return &OCPDNSNameResolver{
+		regularDNSInfo:   make(map[string]namespaceDNSInfo),
+		wildcardDNSInfo:  make(map[string]namespaceDNSInfo),
+		namespaces:       make(map[string]struct{}),
+		minimumTTL:       defaultMinTTL,
+		failureThreshold: defaultFailureThreshold,
+	}
+}
+
+const (
+	// defaultResyncPeriod gives the resync period used for creating the DNSNameResolver informer.
+	defaultResyncPeriod = 24 * time.Hour
+	// defaultMinTTL will be used when minTTL is not explicitly configured.
+	defaultMinTTL int32 = 5
+	// defaultFailureThreshold will be used when failureThreshold is not explicitly configured.
+	defaultFailureThreshold int32 = 5
+)
+
+// initInformer initializes the DNSNameResolver informer.
+func (resolver *OCPDNSNameResolver) initInformer(createClient func() (networkclient.Interface, error), send func(*dnsv1alpha1.DNSNameResolver)) (err error) {
+	// Create the network client.
+	networkClient, err := createClient()
+	if err != nil {
+		return err
+	}
+
+	// Get the client for version v1alpha1 for DNSNameResolver objects.
+	resolver.dnsNameResolverClient = networkClient.NetworkV1alpha1()
```

Comment:
The field name doesn't quite correspond to the value.
```suggestion
	resolver.networkClient = networkClient.NetworkV1alpha1()
```

Reply:
I have changed the name to `ocpNetworkClient`. Similar naming is being followed in ovnk repo.

Reply:
Btw I'm struggling to see where this field is used.

Reply:
This field is used in `handler.go` file for updating the status of the `DNSNameResolver` objects.